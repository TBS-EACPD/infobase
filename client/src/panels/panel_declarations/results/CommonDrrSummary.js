import classNames from "classnames";
import _ from "lodash";
import React, { Fragment } from "react";

import {
  DisplayTable,
  create_text_maker_component,
} from "src/components/index";

import { is_a11y_mode } from "src/core/injected_build_constants";

import { IconArray } from "src/charts/IconArray";
import { StandardLegend } from "src/charts/legends/index";
import { WrappedNivoPie } from "src/charts/wrapped_nivo/index";

import { toggle_list } from "src/general_utils";

import { large_status_icons } from "./result_components";
import {
  ordered_status_keys,
  filter_and_genericize_doc_counts,
  result_statuses,
  result_color_scale,
  get_year_for_doc_key,
} from "./results_common";

import text from "./CommonDrrSummary.yaml";

const { text_maker, TM } = create_text_maker_component(text);

import "./CommonDrrSummary.scss";

const grid_colors = {
  met: "results-icon-array-pass",
  not_met: "results-icon-array-fail",
  not_available: "results-icon-array-na",
  future: "results-icon-array-neutral",
};

const icon_order = _.chain(ordered_status_keys)
  .map((status_key, ix) => [status_key, ix * 5])
  .fromPairs()
  .value();

const MiniLegend = ({ items }) => (
  <div
    style={{
      display: "flex",
      justifyContent: "flex-start",
      flexWrap: "wrap",
      fontSize: "0.87em",
      marginBottom: "5px",
    }}
  >
    {_.map(items, ({ label, id, className }) => (
      <div
        key={id}
        style={{
          display: "flex",
          justifyContent: "flex-start",
          alignItems: "center",
          margin: "5px 15px 5px 0",
        }}
      >
        <div
          style={{
            width: "20px",
            height: "20px",
            marginRight: "5px",
          }}
          className={className}
          title={result_statuses[id].text}
        />
        <span>{label}</span>
      </div>
    ))}
  </div>
);

// TODO: this could be a reusable chart, consider breaking it out at some point
const StatusGrid = ({ met, not_met, not_available, future, drr_key }) => {
  const max_size = 800;

  const counts = { met, not_met, not_available, future };

  const total = met + not_met + not_available + future;
  const shouldFactorDown = total > max_size;
  const icon_array_size_class = classNames(
    "IconArrayItem",
    total > 200 && "IconArrayItem__Small",
    total < 100 && "IconArrayItem__Large"
  );

  const is_single_indicator = _.some(counts, (value) => value === total);

  const data = _.chain(counts)
    .pickBy((val) => val > 0 || is_single_indicator)
    .toPairs()
    .groupBy(([key, _val]) => key)
    .map((amounts, status_key) => {
      const key_total = _.sumBy(amounts, 1);
      return {
        status_key,
        viz_count: shouldFactorDown
          ? Math.ceil((key_total / total) * max_size)
          : key_total,
        real_count: key_total,
      };
    })
    .value();

  const viz_data = _.chain(data)
    .sortBy(({ status_key }) => icon_order[status_key])
    .flatMap(({ viz_count, status_key }) => {
      return _.times(viz_count, () => ({ status_key }));
    })
    .value();

  const legend_data = _.chain(data)
    .map(({ status_key }) => ({
      className: grid_colors[status_key],
      id: status_key,
      label: text_maker(status_key),
      order: icon_order[status_key],
    }))
    .sortBy("order")
    .value();

  if (is_a11y_mode) {
    const a11y_data = _.map(data, ({ status_key, real_count }) => ({
      label: text_maker(status_key),
      real_count,
    }));
    return (
      <DisplayTable
        data={a11y_data}
        column_configs={{
          label: {
            index: 0,
            header: text_maker("status"),
          },
          real_count: {
            index: 1,
            formatter: "big_int",
            header: text_maker("results_icon_array_title", {
              year: get_year_for_doc_key(drr_key),
            }),
          },
        }}
      />
    );
  }

  return (
    <div>
      <div>
        <MiniLegend items={legend_data} />
        <div style={{ marginTop: "3rem" }}>
          {_.chain(viz_data)
            .groupBy("status_key")
            .map((group, status_key) => [group, status_key])
            .sortBy(([_group, status_key]) => icon_order[status_key])
            .map(([group, status_key]) => (
              <IconArray
                key={status_key}
                items={group}
                render_item={({ status_key }) => (
                  <div
                    className={classNames(
                      icon_array_size_class,
                      grid_colors[status_key]
                    )}
                    title={result_statuses[status_key].text}
                  />
                )}
              />
            ))
            .value()}
        </div>
      </div>
    </div>
  );
};

class PercentageViz extends React.Component {
  constructor(props) {
    super(props);
    const { counts } = this.props;

    const all_ids = _.keys(counts);
    const present_ids = _.chain(counts)
      .toPairs()
      .reject((count) => count[1] === 0)
      .fromPairs()
      .value();

    const default_selected =
      _.reject(present_ids, (value, key) => key === "future").length > 0 &&
      !is_a11y_mode
        ? _.without(all_ids, "future")
        : all_ids;

    this.state = {
      selected: default_selected,
    };
  }

  render() {
    const { counts, drr_key } = this.props;
    const { selected } = this.state;

    const all_data = _.chain(counts)
      .map((value, key) => ({
        label: result_statuses[key].text,
        id: key,
        value,
      }))
      .filter((status) => status.value > 0)
      .value();

    const all_data_total = _.sumBy(all_data, "value");

    const graph_data = _.filter(all_data, ({ id }) => _.includes(selected, id));

    const graph_total = _.sumBy(graph_data, "value");
    const colors = ({ id }) => result_color_scale(id);

    const new_summary_text_args = {
      year: get_year_for_doc_key(drr_key),
      drr_subset: graph_total,
      drr_total: all_data_total,
      drr_indicators_met: _.includes(selected, "met") && counts.met,
      drr_indicators_not_met: _.includes(selected, "not_met") && counts.not_met,
      drr_indicators_not_available:
        _.includes(selected, "not_available") && counts.not_available,
      drr_indicators_future: _.includes(selected, "future") && counts.future,
    };

    return (
      <Fragment>
        <div className="row">
          <div className="col-12 col-lg-1 " />
          <div
            className="col-12 col-lg-4 "
            style={{
              display: "flex",
              flexDirection: "column",
              justifyContent: "space-evenly",
            }}
          >
            {!is_a11y_mode && (
              <StandardLegend
                legendListProps={{
                  items: _.chain(all_data)
                    .map(({ label, id }) => ({
                      label: label,
                      active: _.includes(selected, id),
                      id,
                      color: result_color_scale(id),
                    }))
                    .value(),
                  onClick: (id) => {
                    !(selected.length === 1 && selected.includes(id)) &&
                      this.setState({
                        selected: toggle_list(selected, id),
                      });
                  },
                }}
              />
            )}
            <div
              className="standard-legend-container"
              style={{ margin: "5px 0px" }}
            >
              <TM k="drr_summary_stats" args={new_summary_text_args} />
            </div>
          </div>
          <div className="col-12 col-lg-1 " />
          <div className="col-12 col-lg-6  medium-panel-text">
            <WrappedNivoPie
              data={graph_data}
              graph_height="300px"
              is_money={false}
              colors={colors}
              margin={{
                top: 30,
                right: 30,
                bottom: 30,
                left: 30,
              }}
              show_legend={false}
            />
          </div>
        </div>
      </Fragment>
    );
  }
}

export const CommonDrrSummary = ({
  subject,
  drr_key,
  counts,
  verbose_counts,
  results_dept_count,
}) => {
  const current_drr_counts_with_generic_keys = filter_and_genericize_doc_counts(
    verbose_counts,
    drr_key
  );

  const summary_text_args = {
    subject,
    results_dept_count: subject.id === "gov" && results_dept_count,
    year: get_year_for_doc_key(drr_key),
    ...current_drr_counts_with_generic_keys,
  };

  return (
    <Fragment>
      <div className="row align-items-center justify-content-lg-between">
        <div className="col-12 medium-panel-text">
          <TM k="drr_summary_text_intro" args={summary_text_args} />
        </div>
      </div>
      <div className="row align-items-center justify-content-lg-between">
        <div className="col-12 col-lg-7  medium-panel-text">
          <div style={{ padding: "10px" }}>
            <TM k="result_status_explanation" />
            <table>
              <tbody>
                {_.map(ordered_status_keys, (status) => (
                  <tr key={status}>
                    <td style={{ padding: "10px" }}>
                      {large_status_icons[status]}
                    </td>
                    <td>
                      <TM k={`result_status_explanation_${status}`} />
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
          </div>
        </div>
        <div className="col-12 col-lg-5 ">
          <div style={{ padding: "30px" }}>
            <StatusGrid {...counts} drr_key={drr_key} />
          </div>
        </div>
      </div>
      <div className="panel-separator" style={{ marginTop: "0px" }} />
      <div className="row align-items-center justify-content-lg-between">
        <div className={"col-12"}>
          <PercentageViz
            summary_text_args={summary_text_args}
            counts={counts}
            drr_key={drr_key}
          />
        </div>
      </div>
    </Fragment>
  );
};
