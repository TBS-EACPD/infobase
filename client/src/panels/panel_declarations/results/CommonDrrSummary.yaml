drr_summary_text_intro:
  transform: [handlebars,markdown]
  en: |
    {{#if results_dept_count}}
      In **{{year}}**, **{{results_dept_count}}** organizations sought to achieve **{{fmt_big_int drr_results}}** {{gl_sidebar_link "Departmental Results" "DR"}}.
      Progress towards meeting these results was measured using **{{fmt_big_int drr_total}}** {{gl_sidebar_link "Departmental Result Indicators" "IND_RESULT"}}, and a {{gl_sidebar_link "Result Status" "RESULTS_STATUS"}} was assigned to each indicator
      based on the measured outcome (actual result) of the indicator's target.
    {{else}}
      In **{{year}}**, {{subject.name}} sought to achieve **{{fmt_big_int drr_results}}** results.
      Progress towards meeting these results was measured using **{{fmt_big_int drr_total}}** indicators, and a {{gl_sidebar_link "Result Status" "RESULTS_STATUS"}} was assigned to each indicator
      based on the measured outcome (actual result) of the indicator's target.
    {{/if}}
  fr: |
    {{#if results_dept_count}}
      En **{{year}}**, **{{results_dept_count}}** organisations ont cherché à obtenir **{{fmt_big_int drr_results}}** {{gl_sidebar_link "résultats ministériels" "DR"}}.
      Les progrès réalisés pour atteindre ces résultats ont été mesurés au moyen de **{{fmt_big_int drr_total}}** {{gl_sidebar_link "indicateurs de résultat ministériel" "IND_RESULT"}}, et un {{gl_sidebar_link "état du résultat" "RESULTS_STATUS"}}
      a été attribué à chaque indicateur correspondant au résultat obtenu pour la cible de l'indicateur.
    {{else}}
      En **{{year}}**, {{subject.name}} a cherché à atteindre **{{fmt_big_int drr_results}}** résultats.
      Les progrès réalisés pour atteindre ces résultats ont été mesurés au moyen de **{{fmt_big_int drr_total}}** indicateurs, et un {{gl_sidebar_link "état du résultat" "RESULTS_STATUS"}}
      a été attribué à chaque indicateur correspondant au résultat obtenu pour la cible de l'indicateur.
    {{/if}}

drr_summary_stats:
  transform: [handlebars,markdown]
  en: |
    {{#if results_dept_count}}
      In **{{year}}**, there was a total of **{{fmt_big_int drr_total}}** {{gl_sidebar_link "Departmental Result Indicators" "IND_RESULT"}}. Of the **{{fmt_big_int drr_subset}}** selected above:
    {{else}}
      In **{{year}}**, there was a total of **{{fmt_big_int drr_total}}** indicators. Of the **{{fmt_big_int drr_subset}}** selected above: 
    {{/if}}
      {{#if drr_indicators_met}}* **{{fmt_big_int drr_indicators_met}}** (**{{fmt_percentage1 (divide drr_indicators_met drr_subset)}}**) {{gl_sidebar_link "met" "RESULTS_MET"}} their target{{/if}}
      {{#if drr_indicators_not_met}}* **{{fmt_big_int drr_indicators_not_met}}** (**{{fmt_percentage1 (divide drr_indicators_not_met drr_subset)}}**) did {{gl_sidebar_link "not meet" "RESULTS_NOT_MET"}} their target{{/if}}
      {{#if drr_indicators_not_available}}* **{{fmt_big_int drr_indicators_not_available}}** (**{{fmt_percentage1 (divide drr_indicators_not_available drr_subset)}}**) have {{gl_sidebar_link "no result available" "RESULTS_NOT_AVAILABLE"}}{{/if}}
      {{#if drr_indicators_future}}* **{{fmt_big_int drr_indicators_future}}** (**{{fmt_percentage1 (divide drr_indicators_future drr_subset)}}**) are {{gl_sidebar_link "to be achieved in the future" "RESULTS_ONGOING"}}{{/if}}
  fr: |
    {{#if drr_total}}
    {{#if results_dept_count}}
      En **{{year}}**, il y avait **{{fmt_big_int drr_total}}** {{gl_sidebar_link "indicateurs de résultat ministériel" "IND_RESULT"}} en total. Sur les **{{fmt_big_int drr_subset}}** susmentionnés:
    {{else}}
      En **{{year}}**, il y avait **{{fmt_big_int drr_total}}** indicateurs en total. Sur les **{{fmt_big_int drr_subset}}** susmentionnés:
    {{/if}}
      {{#if drr_indicators_met}}* **{{fmt_big_int drr_indicators_met}}** (**{{fmt_percentage1 (divide drr_indicators_met drr_subset)}}**) {{gl_sidebar_link "ont atteint" "RESULTS_MET"}} leur cible{{/if}}
      {{#if drr_indicators_not_met}}* **{{fmt_big_int drr_indicators_not_met}}** (**{{fmt_percentage1 (divide drr_indicators_not_met drr_subset)}}**) {{gl_sidebar_link "n’ont pas atteint" "RESULTS_NOT_MET"}} leur cible{{/if}}
      {{#if drr_indicators_not_available}}* **{{fmt_big_int drr_indicators_not_available}}** (**{{fmt_percentage1 (divide drr_indicators_not_available drr_subset)}}**) {{gl_sidebar_link "n’ont pas de résultat disponible" "RESULTS_NOT_AVAILABLE"}}{{/if}}
      {{#if drr_indicators_future}}* **{{fmt_big_int drr_indicators_future}}** (**{{fmt_percentage1 (divide drr_indicators_future drr_subset)}}**) sont {{gl_sidebar_link "à atteindre dans le futur" "RESULTS_ONGOING"}}{{/if}}
    {{/if}}


results_icon_array_title:
  transform: [handlebars]
  en: |
    {{year}} Performance Indicators
  fr: |
    Indicateurs de performance {{year}}

graph_legend_instructions:
  en: |
    Click below to change which results are counted in the full set:
  fr: |
    Cliquez ci-dessous pour changer l’état des résultats pris en compte dans l’ensemble du cercle


result_status_explanation:
  transform: [handlebars]
  en: The result status depends on how the target was defined, e.g. a minimum target requires a measured outcome (actual result) greater than the target.

  fr: L'état du résultat dépend de la manière dont la cible a été définie, par exemple une cible minimale exige un résultat mesuré (résultat réel) supérieur à la cible.

result_status_explanation_met:
  transform: [handlebars]
  en: If the measured outcome satisfies the target, the indicator is assigned a status of {{gl_sidebar_link "Target met" "RESULTS_MET"}}.
  fr: Si le résultat mesuré correspond à la cible, l'indicateur se voit attribuer l'état {{gl_sidebar_link "Cible atteinte" "RESULTS_MET"}}.

result_status_explanation_not_met:
  transform: [handlebars]
  en: If the measured outcome is outside the range or standard established by the target, the indicator's status is {{gl_sidebar_link "Target not met" "RESULTS_NOT_MET"}}.
  fr: Si le résultat mesuré se situe en dehors de la fourchette ou de la norme établie par la cible, l’état de l'indicateur est {{gl_sidebar_link "Cible non atteinte" "RESULTS_NOT_MET"}}.

result_status_explanation_not_available:
  transform: [handlebars]
  en: Indicators for which no target was set, or for which data was unavailable, are assigned the status {{gl_sidebar_link "No result available" "RESULTS_NOT_AVAILABLE"}}.
  fr: Les indicateurs pour lesquels aucune cible n'a été fixée, ou pour lesquels les données n'étaient pas disponibles, se voient attribuer l’état {{gl_sidebar_link "Résultat non disponible" "RESULTS_NOT_AVAILABLE"}}.

result_status_explanation_future:
  transform: [handlebars]
  en: If the date to achieve the target is after the end of the fiscal year (March 31), or if there is no date to achieve the target, the status is {{gl_sidebar_link "Result to be achieved in the future" "RESULTS_ONGOING"}}.
  fr: Si la date d’atteinte de la cible est postérieure à la fin de l'exercice financier (31 mars), ou s'il n'y a pas de date d’atteinte de la cible, l’état est {{gl_sidebar_link "Résultat à atteindre dans le futur" "RESULTS_ONGOING"}}.