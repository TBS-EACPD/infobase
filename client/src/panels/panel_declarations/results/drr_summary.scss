@import "src/common_style_variables/common-variables";

.IconArrayItem {
  width: 15px;
  height: 15px;
  margin: 0.5px;

  &.IconArrayItem__Small {
    width: 10px;
    height: 10px;
  }

  &.IconArrayItem__Large {
    width: 20px;
    height: 20px;
  }
}

.results-icon-array-pass {
  background-color: #206bbd;
}

.results-icon-array-fail {
  background-color: #4abbc4;
}

.results-icon-array-na {
  background-color: #e89a40;
}

.results-icon-array-neutral {
  background-color: #919bd4;
}
