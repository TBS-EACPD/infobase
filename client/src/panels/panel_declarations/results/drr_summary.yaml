drr_summary_title:
  transform: [handlebars]
  en: |
    Departmental Results Report Summary ({{first_year}}{{#if last_year}} to {{last_year}}{{/if}})
  fr: |
    Sommaire des rapport sur les résultats ministériels ({{first_year}}{{#if last_year}} à {{last_year}}{{/if}})