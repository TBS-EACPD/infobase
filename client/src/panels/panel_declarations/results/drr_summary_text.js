import { create_text_maker_component } from "src/components/index.js";

import text from "./drr_summary_text.yaml";

export const { text_maker, TM } = create_text_maker_component(text);
