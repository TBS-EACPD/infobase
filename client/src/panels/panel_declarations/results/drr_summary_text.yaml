drr_summary_title:
  transform: [handlebars]
  en: |
    {{year}} Department Results Report Summary
  fr: |
    Sommaire des rapports sur les résultats ministériels de {{year}}


# this a multi-level text 
drr_summary_text_intro:
  transform: [handlebars,markdown]
  en: |
    {{#if is_gov}}
      In **{{year}}**, **{{num_depts}}** organizations sought to achieve **{{fmt_big_int drr_results}}** results.
    {{else}}
      In **{{year}}**, {{subj_name subject}} sought to achieve **{{fmt_big_int drr_results}}** results.
    {{/if}}
    Progress towards meeting these results was measured using **{{fmt_big_int drr_total}}** indicators, and a {{gl_tt "Result Status" "RESULTS_STATUS"}} was assigned to each indicator
    based on the measured outcome (actual result) of the indicator's target.
  fr: |
    {{#if is_gov}}
      En **{{year}}**, **{{num_depts}}** organisations ont cherché à obtenir **{{fmt_big_int drr_results}}** résultats.
    {{else}}
      En **{{year}}**, {{subj_name subject}} a cherché à atteindre **{{fmt_big_int drr_results}}** résultats.
    {{/if}}
    Les progrès réalisés pour atteindre ces résultats ont été mesurés au moyen de **{{fmt_big_int drr_total}}** indicateurs, et un {{gl_tt "état du résultat" "RESULTS_STATUS"}}
    a été attribué à chaque indicateur correspondant au résultat obtenu pour la cible de l'indicateur.


new_drr_summary_text_summary:
  transform: [handlebars,markdown]
  en: |
    Of the **{{fmt_big_int drr_total}}** indicators reported:
      {{#if drr_indicators_met}}* **{{fmt_big_int drr_indicators_met}}** (**{{fmt_percentage1 (divide drr_indicators_met drr_total)}}**) {{gl_tt "met" "RESULTS_MET"}} their target{{/if}}
      {{#if drr_indicators_not_met}}* **{{fmt_big_int drr_indicators_not_met}}** (**{{fmt_percentage1 (divide drr_indicators_not_met drr_total)}}**) did {{gl_tt "not meet" "RESULTS_NOT_MET"}} their target{{/if}}
      {{#if drr_indicators_not_available}}* **{{fmt_big_int drr_indicators_not_available}}** (**{{fmt_percentage1 (divide drr_indicators_not_available drr_total)}}**) have {{gl_tt "no result available" "RESULTS_NOT_AVAILABLE"}}{{/if}}
      {{#if drr_indicators_future}}* **{{fmt_big_int drr_indicators_future}}** (**{{fmt_percentage1 (divide drr_indicators_future drr_total)}}**) are {{gl_tt "to be achieved in the future" "RESULTS_ONGOING"}}{{/if}}
  fr: |
    {{#if drr_total}}
    De ces **{{fmt_big_int drr_past_total}}** indicateurs rapportés:
      {{#if drr_indicators_met}}* **{{fmt_big_int drr_indicators_met}}** (**{{fmt_percentage1 (divide drr_indicators_met drr_total)}}**) {{gl_tt "ont atteint" "RESULTS_MET"}} leur cible{{/if}}
      {{#if drr_indicators_not_met}}* **{{fmt_big_int drr_indicators_not_met}}** (**{{fmt_percentage1 (divide drr_indicators_not_met drr_total)}}**) {{gl_tt "n’ont pas atteint" "RESULTS_NOT_MET"}} leur cible{{/if}}
      {{#if drr_indicators_not_available}}* **{{fmt_big_int drr_indicators_not_available}}** (**{{fmt_percentage1 (divide drr_indicators_not_available drr_total)}}**) {{gl_tt "n’ont pas de résultat disponible" "RESULTS_NOT_AVAILABLE"}}{{/if}}
      {{#if drr_indicators_future}}* **{{fmt_big_int drr_indicators_future}}** (**{{fmt_percentage1 (divide drr_indicators_future drr_total)}}**) sont {{gl_tt "à atteindre dans le future" "RESULTS_ONGOING"}}{{/if}}
    {{/if}}

gov_drr_summary_org_table_text:
  transform: [handlebars,markdown]
  en: |
    To learn more on the result and their associated performance indicators for an organization, 
    simply click on the name of the organization you want to explore to be redirected toward their specific results infographic.
    
  fr: |
    Pour en savoir plus sur les résultats et les indicateurs de rendement d'une organisation, 
    il suffit de cliquer sur le nom de l'organisation pour être dirigé vers l'infographie où figurent ses résultats particuliers.

results_icon_array_title:
  transform: [handlebars]
  en: |
    {{year}} Performance Indicators
  fr: |
    Indicateurs de performance {{year}}

graph_legend_instructions:
  en: |
    Click to change which results are counted in the full set:
  fr: |
    Cliquez ci-dessous pour changer l’état des résultats pris en compte dans l’ensemble du cercle


result_status_explanation:
  transform: [handlebars,markdown]
  en: |
    The result status depends on how the target was defined, e.g. a minimum target requires a measured outcome greater than the target.
    
    * If the measured outcome satisfies the target, the indicator is assigned a status of {{gl_tt "Target met" "RESULTS_MET"}}.
    * If the measured outcome is outside the range or standard established by the target, the indicator's status is {{gl_tt "Target not met" "RESULTS_NOT_MET"}}.
    * Indicators for which no target was set, or for which data was unavailable, are assigned the status {{gl_tt "No result available" "RESULTS_NOT_AVAILABLE"}}.
    * If the date to achieve the target is after the end of the fiscal year (March 31), or if there is no date to achieve the target, the status is {{gl_tt "Result to be achieved in the future" "RESULTS_ONGOING"}}.
  fr: |
    L'état du résultat dépend de la manière dont la cible a été définie, par exemple une cible minimale exige un résultat mesuré (résultat réel) supérieur à la cible.
    
    * Si le résultat mesuré correspond à la cible, l'indicateur se voit attribuer l'état {{gl_tt "Cible atteinte" "RESULTS_MET"}}.
    * Si le résultat mesuré se situe en dehors de la fourchette ou de la norme établie par la cible, l’état de l'indicateur est {{gl_tt "Cible non atteinte" "RESULTS_NOT_MET"}}.
    * Les indicateurs pour lesquels aucune cible n'a été fixée, ou pour lesquels les données n'étaient pas disponibles, se voient attribuer l’état {{gl_tt "Résultat non disponible" "RESULTS_NOT_AVAILABLE"}}.
    * Si la date d’atteinte de la cible est postérieure à la fin de l'exercice financier (31 mars), ou s'il n'y a pas de date d’atteinte de la cible, l’état est {{gl_tt "Résultat à atteindre dans le futur" "RESULTS_ONGOING"}}.
