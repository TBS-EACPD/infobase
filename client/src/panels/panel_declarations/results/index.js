export { declare_explore_results_panel } from "./result_drilldown/result_drilldown";
export { declare_results_table_panel } from "./result_flat_table";
export { declare_drr_summary_panel } from "./drr_summary";
export { declare_gov_drr_panel } from "./gov_drr";
export { declare_gov_dp_panel } from "./gov_dp";
export { declare_results_intro_panel } from "./results_intro";
