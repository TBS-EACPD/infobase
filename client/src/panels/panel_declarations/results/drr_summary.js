import _ from "lodash";
import React, { useState } from "react";

import { InfographicPanel } from "src/panels/panel_declarations/InfographicPanel";
import { declare_panel } from "src/panels/PanelRegistry";

import {
  create_text_maker_component,
  Tabs,
  LeafSpinner,
} from "src/components/index";

import { CommonDrrSummary } from "./CommonDrrSummary";

import {
  ResultCounts,
  GranularResultCounts,
  get_year_for_doc_key,
  load_results_api,
  hierarchy_to_counts,
} from "./results_common";

import text from "./drr_summary.yaml";

const { text_maker } = create_text_maker_component(text);

const get_verbose_counts = (subject) => {
  switch (subject.subject_type) {
    case "dept":
      return ResultCounts.get_dept_counts(subject.id);
    case "crso":
      return _.chain([subject.id, ..._.map(subject.programs, "id")])
        .map((id) => GranularResultCounts.get_subject_counts(id))
        .reduce(
          (accumulator, counts) => _.mergeWith(accumulator, counts, _.add),
          {}
        )
        .value();
    case "program":
      return GranularResultCounts.get_subject_counts(subject.id);
  }
};
const get_drr_keys_with_data = (subject) =>
  _.chain(get_verbose_counts(subject))
    .pickBy(
      (count_value, count_key) =>
        /^drr[0-9]*_total$/.test(count_key) && count_value !== 0
    )
    .keys()
    .map((count_key) => _.split(count_key, "_")[0])
    .value();

const DrrSummary = ({ subject, drr_keys, verbose_counts }) => {
  const [drr_key, set_drr_key] = useState(_.last(drr_keys));

  const { loading, data } = load_results_api(subject);

  if (loading) {
    return <LeafSpinner config_name="subroute" />;
  }

  const counts = hierarchy_to_counts(data, drr_key);

  const summary = (
    <CommonDrrSummary
      subject={subject}
      drr_key={drr_key}
      verbose_counts={verbose_counts}
      counts={counts}
    />
  );

  if (drr_keys.length > 1) {
    return (
      <Tabs
        tabs={_.chain(drr_keys)
          .map((drr_key) => [drr_key, get_year_for_doc_key(drr_key)])
          .fromPairs()
          .value()}
        open_tab_key={drr_key}
        tab_open_callback={set_drr_key}
      >
        {summary}
      </Tabs>
    );
  } else {
    return summary;
  }
};

export const declare_drr_summary_panel = () =>
  declare_panel({
    panel_key: "drr_summary",
    subject_types: ["dept", "crso", "program"],
    panel_config_func: (subject_type) => ({
      legacy_non_table_dependencies:
        subject_type === "dept"
          ? ["requires_result_counts"]
          : ["requires_granular_result_counts"],
      get_title: ({ subject }) => {
        const drr_keys = get_drr_keys_with_data(subject);

        return (
          !_.isEmpty(drr_keys) &&
          text_maker("drr_summary_title", {
            first_year: get_year_for_doc_key(_.first(drr_keys)),
            last_year:
              drr_keys.length > 1 && get_year_for_doc_key(_.last(drr_keys)),
          })
        );
      },
      get_dataset_keys: () => ["actual_results"],
      calculate: ({ subject }) => {
        const drr_keys = get_drr_keys_with_data(subject);

        if (_.isEmpty(drr_keys)) {
          return false;
        }

        return {
          drr_keys,
          verbose_counts: get_verbose_counts(subject),
        };
      },
      render: ({
        title,
        subject,
        calculations,
        footnotes,
        sources,
        datasets,
      }) => {
        const { verbose_counts, drr_keys } = calculations;

        return (
          <InfographicPanel {...{ title, footnotes, sources, datasets }}>
            <DrrSummary
              subject={subject}
              drr_keys={drr_keys}
              verbose_counts={verbose_counts}
            />
          </InfographicPanel>
        );
      },
    }),
  });
