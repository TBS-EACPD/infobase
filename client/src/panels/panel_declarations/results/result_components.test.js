import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

import { Indicator } from '../../models/results.js';
import {
  IndicatorDisplay,
  QuadrantDefList,
  status_icons,
  indicators_period_span_str,
  StatusIconTable,
  InlineStatusIconList,
  HorizontalStatusTable,
} from './result_components.js';


let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});


describe("result_components",()=>{
  it("renders correctly",()=>{


    const ind = {
      "id": "NI001-0030-2017",
      "stable_id": "NI001-0030-2017",
      "result_id": "NR001-0025-2017",
      "name": "The current year producersâ€™ net market income plus Business Risk Management program payments as a percent of the previous five year average",
      "doc": "drr17",
      "target_year": 2018,
      "target_month": 3,
      "is_reporting_discontinued": false,
      "target_type": "percent",
      "target_min": "85",
      "target_max": null,
      "target_narrative": null,
      "measure": null,
      "previous_year_target_type": null,
      "previous_year_target_min": null,
      "previous_year_target_max": null,
      "previous_year_target_narrative": null,
      "previous_year_measure": null,
      "target_explanation": "Strong sector income growth has lead to this indicator exceeding its targets.",
      "result_explanation": null,
      "actual_datatype": "percent",
      "actual_result": "121",
      "status_key": "met",
      "methodology": null
    };

    act(() => {
      render(
        <IndicatorDisplay indicators={[ind]} />,
        container
      );
    });
    
    debugger;
    expect(container.textContent).toMatchSnapshot();
  });

});