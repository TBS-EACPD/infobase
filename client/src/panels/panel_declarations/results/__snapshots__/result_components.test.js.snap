// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`result_components renders correctly 1`] = `
"IndicatorThe current year producersâ€™ net market income plus Business Risk Management program payments as a percent of the previous five year averageNewReport2019-20 Departmental Results ReportDate to achieve targetTargetTarget not specifiedActual result121%StatusMetTarget metTarget explanationStrong sector income growth has lead to this indicator exceeding its targets.
"
`;
